### PR TITLE
 Create team endpoint

### PR DESCRIPTION
**Pull request by eduardavercosa**
## Objective and Motivation
- Create team endpoint

### CODE REVIEW

File: backend/battling/api/endpoints.py [hunk 1]
```diff
@@ -31,3 +31,12 @@ def get_queryset(self):
 class CreateBattle(generics.CreateAPIView):
     serializer_class = BattleSerializer
     permission_classes = [permissions.IsAuthenticatedOrReadOnly]
+
+
+class CreateTeam(generics.UpdateAPIView):
+    serializer_class = BattleTeamSerializer
+    permission_classes = [permissions.IsAuthenticatedOrReadOnly]
```
Comment (rsarai) [inline backend/battling/api/endpoints.py]: Should we use this permission?

File: backend/battling/forms.py [hunk 2]
```diff
@@ -7,9 +7,9 @@
 from pokemon.helpers import (
     get_or_create_pokemon,
     get_pokemon_from_api,
-    has_repeated_pokemon,
+    has_repeated_pokemon_old,
```
Comment (rsarai) [inline backend/battling/forms.py]: why was this renamed? We should not add suffix with time constraints our code, this makes the code dated and hard to understand

File: backend/pokemon/helpers.py [hunk 3]
```diff
@@ -68,9 +68,23 @@ def has_repeated_positions(pokemon_position_list):
...
-def has_repeated_pokemon(pokemon_names):
+def has_repeated_pokemon_old(pokemon_names):
     for pokemon in pokemon_names:
         pokemon_count = pokemon_names.count(pokemon)
         if pokemon_count > 1:
             return True
     return False
+
+
+def has_repeated_pokemon(pokemons_list):
```
Comment (rsarai) [inline backend/pokemon/helpers.py:64]: We can't have two functions doing the same thing. In situations like this, we should refactor the multiple places of the code to use the same function.

File: backend/battling/api/endpoints.py [hunk 4]
```diff
@@ -31,3 +31,12 @@ def get_queryset(self):
 class CreateBattle(generics.CreateAPIView):
     serializer_class = BattleSerializer
     permission_classes = [permissions.IsAuthenticatedOrReadOnly]
+
+
+class CreateTeam(generics.UpdateAPIView):
+    serializer_class = BattleTeamSerializer
```
Comment (rsarai) [inline backend/battling/api/endpoints.py]: Why the `CreateTeam` endpoint is using a `BattleTeamSerializer`? The names here could be improved

File: backend/battling/api/serializers.py [hunk 5]
```diff
@@ -20,17 +22,26 @@ class Meta:
...
+    class Meta:
+        model = PokemonTeam
+        fields = ("id", "teams", "pokemon", "order")
 
 
 class BattleSerializer(serializers.ModelSerializer):
     creator = UserSerializer(read_only=True)
     opponent = UserSerializer(read_only=True)
     teams = TeamSerializer(many=True, read_only=True)
+    pokemon_teams = PokemonTeamSerializer(many=True, required=False)
```
Comment (rsarai) [inline backend/battling/api/serializers.py]: I don't see the utility of this attribute, where are you thinking about using this?

File: backend/battling/api/serializers.py [hunk 6]
```diff
@@ -20,17 +22,26 @@ class Meta:
...
     trainer = UserSerializer()
-    pokemons = PokemonSerializer(many=True)
 
     class Meta:
         model = Team
-        fields = ("id", "battle", "trainer", "pokemons")
+        fields = ("id", "battle", "trainer")
+
+
+class PokemonTeamSerializer(serializers.ModelSerializer):
```
Comment (rsarai) [inline backend/battling/api/serializers.py]: I also don't see the utility of this serializer, what are your plans for this?

File: backend/battling/api/urls.py [hunk 7]
```diff
@@ -7,4 +7,5 @@
     path("battles/list/", endpoints.BattleList.as_view(), name="battle-list"),
     path("battles/<int:pk>/", endpoints.BattleDetail.as_view(), name="battle-detail"),
     path("battles/create/", endpoints.CreateBattle.as_view(), name="create-battle"),
+    path("battles/team/<int:pk>/edit/", endpoints.CreateTeam.as_view(), name="create-team"),
```
Comment (rsarai) [inline backend/battling/api/urls.py]: There's no need for the prefix `battle` here, `team/<int:pk>/edit/` fits better

File: backend/battling/api/serializers.py [hunk 8]
```diff
@@ -66,3 +79,80 @@ def create(self, validated_data):
 
         set_up_battle_teams_and_send_invite_email(instance)
         return instance
+
+
+class BattleTeamSerializer(serializers.ModelSerializer):
```
Comment (rsarai) [inline backend/battling/api/serializers.py]: I don't see a problem with having separated `read` and `write` serializers. But they need to be named accordingly and have the logic splitted. What was the reason for this change?

File: backend/battling/api/serializers.py [hunk 9]
```diff
@@ -66,3 +79,80 @@ def create(self, validated_data):
...
+    team = TeamSerializer(read_only=True)
```
Comment (rsarai) [inline backend/battling/api/serializers.py]: This is a read-only field and should not be used on a write-only serializer

File: backend/battling/api/serializers.py [hunk 10]
```diff
@@ -66,3 +79,80 @@ def create(self, validated_data):
...
+    pokemons = PokemonSerializer(many=True, read_only=True)
```
Comment (rsarai) [inline backend/battling/api/serializers.py]: Same thing here, this is a read-only field and should not be used on a write-only serializer

File: backend/battling/api/serializers.py [hunk 11]
```diff
@@ -66,3 +79,80 @@ def create(self, validated_data):
...
+        creator_pokemons = creator.pokemons.all()
+
+        opponent = (
+            Team.objects.filter(battle=battle, trainer=battle.opponent)
+            .prefetch_related("pokemons")
+            .get()
+        )
+        opponent_pokemons = opponent.pokemons.all()
+
+        return creator_pokemons and opponent_pokemons
```
Comment (rsarai) [inline backend/battling/api/serializers.py]: This method should be in a separated place, since it can be used on the serializer and on the form

**Review by rsarai — CHANGES_REQUESTED** [review]

File: backend/battling/api/endpoints.py (hunk 4 above)
Comment (eduardavercosa) [reply backend/battling/api/endpoints.py]: Because the BattleTeamSerializer is where I create the team.... but I'll change the names

**Review by rsarai — APPROVED** [review]

**Pull request merged**